### ISSUE TITLE
noetic process has died

### ISSUE THREAD

**Issue by Panda12580** [pal-robotics/aruco_ros#133]
Hello, may I ask why when I run single,launch file, call rqt to see the result will kill the process? I have checked the set parameters and there is no error. I ask for your help, thank you.

![Screenshot from 2024-04-20 19-00-47](https://github.com/pal-robotics/aruco_ros/assets/124687326/a338d7a0-1436-4565-90d0-534b1fa0773b)

**Comment by saikishor** [comment]
Hello @Panda12580!

It is indeed weird that setting parameters through service works but not through the rqt reconfigure, at the end of the day, they are doing the same. We cannot see the screenshot you have attached. Could you reattach it?

Thank you

**Comment by Panda12580** [comment]
@saikishor Thank you for your reply. I have updated and uploaded the picture. Please check where my mistake is.

**Comment by Panda12580** [comment]
I set up the single.launch file and then start rs_camera.launch, but I want to see images via rqt and subscribed to aruco_ros/result , 
 the background process will be killed as soon as I do this.

**Comment by saikishor** [comment]
@Panda12580 I don't see the relation between rqt image viewer and the point you said about setting parameters through the command line. They are completely different.

What I recommend is to try to subscribe to the topic in the command line and see what would happen. This part is something that is tested, so I expect it to work without issues.

Thank you,

Best Regards,
PAL Robotics Team

**Comment by Oscar-qi** [comment]
![aruco-single-node-died](https://github.com/pal-robotics/aruco_ros/assets/54581708/eca01bee-e353-4474-82b9-1ce7029aecae)
@saikishor Hello, I have been using the aruco-ros package recently on the Ubuntu 20.04 system. When I run the following command:

bash
rosrun image_view image_view image:=/aruco_single/result
the single.launch node dies.

**Comment by saikishor** [comment]
Can you let us know what happens if you subscribe through command line?

**Comment by Oscar-qi** [comment]
I opened three terminals and ran the driver node for the d435 camera, the single_realsense node for aruco_ros, and subscribed to aruco_single/result using the command line。

> The terminal display for subscribing to aruco_single/result using the command line reads:
rosrun image_view image_viewer  image:=/aruco_single/result
[INFO] 【1713951282.914114320]: Inttlaltzing nodelet wtth 4 worker threads.
[INFO][1713951283.689522413]: Using transport "raw"

> When I subscribe to the aruco_single/result topic, it shows the following in the terminal of aruco_ros:

NODES
aruco_single (aruco_ros/stngle)
ROS_MASTER_URI=http://Localhost:11311
process[aruco_stngle-1]:started with pid[80918]
[ WARN] [1713951115.318433478]:Corner refinenent opttons have been removed in Aruco 3.0.0, corner_refinenent ROS paraneter ts deprecated
[ WARN] [1713951116.014256637]: normaltzeImagerllumtnatton is untnplemented!
[aruco single-1] process has dled [pid 80918, extt code -l1, cmd /home/rm/catkin _ws/devel/lib/aruco_ros/single /camera_info:=/camera/color/canera_info   /image:=/camera/color/image_raw __name:=aruco_single __log:=/home/rm/.ros/log/451bc2ea-021d-11ef-8aea-48b02d15ce9b/aruco single-1.log].
log file：/home/rm/.ros/log/451bc2ea-021d-11ef-Baea-48bo2d15ce9b/aruco single-1*
.1og
all processes on machine have died, roslaunch will exit shuttting down processing monitor...
... shutting down processing monitor complete
done

**Comment by Oscar-qi** [comment]
> Can you let us know what happens if you subscribe through command line?

I opened three terminals and ran the driver node for the d435 camera, the single_realsense node for aruco_ros, and subscribed to aruco_single/result using the command line。

The terminal display for subscribing to aruco_single/result using the command line reads:
rosrun image_view image_viewer image:=/aruco_single/result
[INFO] 【1713951282.914114320]: Inttlaltzing nodelet wtth 4 worker threads.
[INFO][1713951283.689522413]: Using transport "raw"

When I subscribe to the aruco_single/result topic, it shows the following in the terminal of aruco_ros:

NODES
aruco_single (aruco_ros/stngle)
ROS_MASTER_URI=http://Localhost:11311
process[aruco_stngle-1]:started with pid[80918]
[ WARN] [1713951115.318433478]:Corner refinenent opttons have been removed in Aruco 3.0.0, corner_refinenent ROS paraneter ts deprecated
[ WARN] [1713951116.014256637]: normaltzeImagerllumtnatton is untnplemented!
[aruco single-1] process has dled [pid 80918, extt code -l1, cmd /home/rm/catkin _ws/devel/lib/aruco_ros/single /camera_info:=/camera/color/canera_info /image:=/camera/color/image_raw __name:=aruco_single __log:=/home/rm/.ros/log/451bc2ea-021d-11ef-8aea-48b02d15ce9b/aruco single-1.log].
log file：/home/rm/.ros/log/451bc2ea-021d-11ef-Baea-48bo2d15ce9b/aruco single-1*
.1og
all processes on machine have died, roslaunch will exit shuttting down processing monitor...
... shutting down processing monitor complete
done

**Comment by chenglei11** [comment]
hi I ran into the same problem with you, have you solved it? could you please tell me how to solve this one? you can reply me in Chinese as well

**Comment by Panda12580** [comment]
@chenglei11 I was able to implement it on another noetic version of the computer, but the code used was previous, that is, I'm not sure whether the author of the package changed the code to cause the conflict or the computer environment itself, so I guess if I can use the earlier version of the noetic package to try it?

**Comment by Panda12580** [comment]
@saikishor I'm sorry for not replying in time. I mean, when I select the topic result from the view-image option in the rqt visual interface, the program running single.launch will be killed, which means that we need to subscribe to the topic result when we recognize the QR code. No matter what the situation is, once we subscribe to the topic, The program will be killed.

**Comment by chenglei11** [comment]
> @chenglei11 I was able to implement it on another noetic version of the computer, but the code used was previous, that is, I'm not sure whether the author of the package changed the code to cause the conflict or the computer environment itself, so I guess if I can use the earlier version of the noetic package to try it?

yeah i can also implement it on another computer, have you tried this method? did it work?

**Comment by BUAAPY** [comment]
> @chenglei11 I was able to implement it on another noetic version of the computer, but the code used was previous, that is, I'm not sure whether the author of the package changed the code to cause the conflict or the computer environment itself, so I guess if I can use the earlier version of the noetic package to try it?

Hey, I've met the same problem. Have you solved the problem? which version have you tried?

**Comment by Panda12580** [comment]
I'm using the noetic branch, and the code I'm using is from earlier, on my other computer. Unfortunately, my rough review of the new code differs greatly from the current one. So I'm not sure what the problem is, the environment configuration problem is very likely.

**Comment by Panda12580** [comment]
@BUAAPY I'm using the noetic branch, and the code I'm using is from earlier, on my other computer. Unfortunately, my rough review of the new code is not very different from the current one. So I'm not sure what the problem is, the environment configuration problem is very likely.

**Comment by LuShuaiqi-Geo** [comment]
I encountered the same problem. My image acquisition operations were the same as the author's. I also used neotic. But the difference is that after I tried to get the image, it did not kill aruco_single, but got stuck. It could not get the image and did not kill the process. I feel very strange and hope there is a solution to this problem

**Comment by BUAAPY** [comment]
I've tried to print messages during the running process and the program has stucked before an opencv function (the resize function). However, the program runs smoothly in another noetic environment. The only difference is the version of opencv (4.5.4&nbsp; in the stucked one, 4.2.0 the other one).&nbsp;


According to other blogs the ROS noetic uses its default opencv (4.2.0), especially its cv_bridge package. The aruco_ros uses my installed opencv (4.5.4) automatically during the cmake process, though the minimum version is set to 4.2.0 in the cmake file. I dont know if this is the issue and how to fix it. Maybe uninstalling the latest opencv and using the default one o noetic may help.



---Original---
From: "Lu ***@***.***&gt;
Date: Tue, Nov 12, 2024 17:01 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [pal-robotics/aruco_ros] noetic process has died (Issue #133)




 
I encountered the same problem. My image acquisition operations were the same as the author's. I also used neotic. But the difference is that after I tried to get the image, it did not kill aruco_single, but got stuck. It could not get the image and did not kill the process. I feel very strange and hope there is a solution to this problem

**Comment by saikishor** [comment]
Exactly, that should be the issue

**Comment by Panda12580** [comment]
@saikishor It doesn't seem right either. I set the opencv version to 4.2 during the compilation process but the problem still occurs. I default to 4.6, then I install 4.2 and change the desired opencv version to 4.2 in the cmakelists file in aruco_ros, and I print out to confirm that I'm using 4.2 in singe.cpp but still get errors. So we sincerely hope that you can help us solve this problem！😊

**Comment by andrewhutong12345** [comment]
@Panda12580 The `aruco_ros` package actually uses `cv_bridge` to find opencv, so you should make sure the cv_bridge is point to the right opencv.
For me, the problem is that when I install the ros-noetic, the version of the default opencv is 4.5.2, not 4.2.0 . I guess maybe the reason is that my platform is JETSON ORIN. So just make sure the opencv' s version is 4.2.0 and cv_bridge is indeed point to that opencv!

**Comment by Panda12580** [comment]
@andrewhutong12345 Yes, I think you make a lot of sense, I just print in the single.cpp file whether the program is running with cv4.2 ignoring the cv_bridge pointing problem, I'll try to see if I can fix it.

**Comment by saikishor** [comment]
Hello! 

Sorry for not being able to respond earlier. I'll see if I can find time to be able to add something to the code, such that if the opencv version is not the expected one. It fails more explicitly with a proper message. 

It would be nice to see the difference between 4.2.0 and 4.5.x and see if we can have it working in both. If someone is interested to take a look, feel free to open a PR. I'll be happy to review and get it merged.

Thank you

**Comment by 307321587** [comment]
> @Panda12580 The `aruco_ros` package actually uses `cv_bridge` to find opencv, so you should make sure the cv_bridge is point to the right opencv. For me, the problem is that when I install the ros-noetic, the version of the default opencv is 4.5.2, not 4.2.0 . I guess maybe the reason is that my platform is JETSON ORIN. So just make sure the opencv' s version is 4.2.0 and cv_bridge is indeed point to that opencv!

This reason is correct. I find that the cv_bridge .cmake file in the path /opt/ros/noetic will set libraries paths such as "libopencv4.2xx". If the local opencv is not 4.2 or the opencv library is "libopencv_world", the program will died without the libraries. 
The solution is simple. I removed the cv_bridge and download its source code in github. https://github.com/ros-perception/vision_opencv. Remember change the branch to noetic. Then put the source code in the workspace with aruco lib and compile them togther, the problem is solved.

```
sudo apt-get remove ros-melodic-cv-bridge
git clone https://github.com/ros-perception/vision_opencv.git
cd vision_opencv
gitcheckout noetic
catkin_make
```

However, when remove cv_bridge, my image_view plugin is also removed. It is needed to install it.
```
sudo apt-get install ros-noetic-rqt-common-plugins
```

These reference wedsites also guided me: 
https://zhuanlan.zhihu.com/p/392939687
https://www.cnblogs.com/gooutlook/p/18394497
